mixer/pkg/protobuf/yaml/dynamic/auth.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2018 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamic
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/clientcredentials"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+
+	policypb "istio.io/api/policy/v1beta1"
+	"istio.io/istio/pkg/log"
+)
+
+const (
+	//
+	publicCACertsFile = "/etc/ssl/certs/ca-certificates.crt"
+)
+
+type authHelper struct {
+	authCfg          *policypb.Authentication
+	skipVerification bool
+}
+
+type headerToken struct {
+	header string
+	prefix string
+	token  []byte
+}
+
+func (c *headerToken) GetRequestMetadata(context.Context, ...string) (map[string]string, error) {
+	if c.prefix == "" {
+		return map[string]string{
+			c.header: string(c.token),
+		}, nil
+	}
+	return map[string]string{
+		c.header: c.prefix + " " + string(c.token),
+	}, nil
+}
+
+func (c *headerToken) RequireTransportSecurity() bool {
+	return true
+}
+
+func loadCerts(caFile string) (*x509.CertPool, error) {
+	// load ca cert.
+	caCertPool := x509.NewCertPool()
+	// load public root cert.
+	publicCerts, err := ioutil.ReadFile(publicCACertsFile)
+	if err != nil {
+		log.Warnf("Cannot load public root cert: %v", err)
+	} else {
+		caCertPool.AppendCertsFromPEM(publicCerts)
+	}
+
+	// load ca file provided by file.
+	if caFile != "" {
+		caCerts, err := ioutil.ReadFile(caFile)
+		if err != nil {
+			return nil, err
+		}
+		caCertPool.AppendCertsFromPEM(caCerts)
+	}
+	return caCertPool, nil
+}
+
+func buildMTLSDialOption(mtlsCfg *policypb.Mutual, skipVerify bool) ([]grpc.DialOption, error) {
+	// load peer cert/key.
+	pk := mtlsCfg.GetPrivateKey()
+	cc := mtlsCfg.GetClientCertificate()
+	peerCert, err := tls.LoadX509KeyPair(cc, pk)
+	if err != nil {
+		return nil, errors.Errorf("load peer cert/key error: %v", err)
+	}
+
+	// Load certificates, include public certs and certs from config.
+	caCertPool, err := loadCerts(mtlsCfg.GetCaCertificates())
+	if err != nil {
+		return nil, fmt.Errorf("failed build tls dial option: ca cert cannot be load %v", err)
+	}
+
+	tc := credentials.NewTLS(&tls.Config{
+		Certificates:       []tls.Certificate{peerCert},
+		RootCAs:            caCertPool,
+		InsecureSkipVerify: skipVerify,
```

Comment:
What about ServerName?
```
        // ServerName indicates the name of the server requested by the client
        // in order to support virtual hosting. ServerName is only set if the
        // client is using SNI (see
        // https://tools.ietf.org/html/rfc4366#section-3.1).
```
This might be useful for SNI (proxy-fronted backend)

Reply:
Good point, unfortunately serverName is not in API. I think most expected deployment would be backend terminating TLS, so it should be fine for 1.1. I will track it as future work in #7637.

Reply:
I created https://github.com/istio/api/pull/789 to add server_name to release 1.1. If it could be checked in, I will add them here. Thanks!